### PR TITLE
Hetzner: initialize instance_id from system-serial-number 

### PR DESCRIPTION
**Pull request by asciiprod**
## Proposed Commit Message
Hetzner Cloud also provides the instance ID in SMBIOS information. Use it to locally check_instance_id and to compared with instance_id from metadata service.

This also prevents reconfiguring an instance in case metadata server is temporarily unavailable or cannot be reached due to broken networking or misconfigured firewall

## Additional Context
* https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/1885527

## Test Steps
* Create an Ubuntu instance on Hetzner Cloud.
* Add changed DS
* Create snapshot from instance
* Create new instance from snapshot should run without errors. cloud-init.log should display instance_id/serial from SMBIOS

## Checklist:
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

**Review by igalic — COMMENTED** [review]
the code looks sensible, I'm just confused by the 26 commits

maybe a rebase would clean this up?

**Comment by igalic** [comment]
that looks much better!

i think i'd you fix flake aaaaand your own tests

```
tests/unittests/test_datasource/test_hetzner.py::TestDataSourceHetzner::test_not_on_hetzner_returns_false FAILED [ 36%]
tests/unittests/test_datasource/test_hetzner.py::TestDataSourceHetzner::test_read_data FAILED [ 36%]
```

we should be good to go

File: cloudinit/sources/DataSourceHetzner.py [hunk 1]
```diff
@@ -3,15 +3,18 @@
 #
 # This file is part of cloud-init. See LICENSE file for license information.
 #
-"""Hetzner Cloud API Documentation.
-   https://docs.hetzner.cloud/"""
+# Hetzner Cloud API:
+# https://docs.hetzner.cloud/
```
Comment (smoser) [inline cloudinit/sources/DataSourceHetzner.py:7]: this change seems unrelated, and changes a docstring to a comment.
On master:

```
>>> from cloudinit.sources import DataSourceHetzner
>>> print(DataSourceHetzner.__doc__)
Hetzner Cloud API Documentation.
   https://docs.hetzner.cloud/
```

But on your branch:
```
>>> from cloudinit.sources import DataSourceHetzner
>>> print(DataSourceHetzner.__doc__)
None
```

File: cloudinit/sources/DataSourceHetzner.py [hunk 2]
```diff
@@ -96,9 +106,20 @@ def network_config(self):
...
+def get_hcloud_data():
+    vendor_name = util.read_dmi_data('system-manufacturer')
+    if vendor_name != "Hetzner":
+        return (False, None)
+
+    serial = util.read_dmi_data("system-serial-number")
+    if serial:
+        LOG.debug("Running on Hetzner Cloud: serial=%s", serial)
+    else:
+        msg = ("Hetzner Cloud detected, but no serial found")
```
Comment (smoser) [inline cloudinit/sources/DataSourceHetzner.py:118]: the () are unnecessary here.

File: cloudinit/sources/DataSourceHetzner.py [hunk 3]
```diff
@@ -96,9 +106,20 @@ def network_config(self):
         return self._network_config
 
 
-def on_hetzner():
-    return util.read_dmi_data('system-manufacturer') == "Hetzner"
+def get_hcloud_data():
+    vendor_name = util.read_dmi_data('system-manufacturer')
+    if vendor_name != "Hetzner":
+        return (False, None)
```
Comment (smoser) [inline cloudinit/sources/DataSourceHetzner.py:112]: Any  reason for getting rid of the simple 'on_hetzner' function?
It provides an easy way with no side effects to tell if you're on hetzner cloud.

Its not a big deal, just curious why you chose to combine it with this function that also reads the serial.

**Review by smoser — CHANGES_REQUESTED** [review]

File: cloudinit/sources/DataSourceHetzner.py (hunk 3 above)
Comment (asciiprod) [reply cloudinit/sources/DataSourceHetzner.py:112]: I think you were the one who added it in the first place (after I submitted our DS) to create aa fast exit. Before making this change, I looked at how various other DS handle this  (e.g Googles platform_reports_gce or DO's get_sysinfo). Just wanted to make it sort of consistent with what is already there

Comment (smoser) [reply cloudinit/sources/DataSourceHetzner.py:112]: If we are ever going to go to a "factory" type of solution, then some simple method in each datasoruce that simply returns boolean (running there or not) woudl be useful.

But we're not there yet.

File: cloudinit/sources/DataSourceHetzner.py [hunk 4]
```diff
@@ -69,14 +75,18 @@ def get_data(self):
         # hostname is name provided by user at launch.  The API enforces it is
         # a valid hostname, but it is not guaranteed to be resolvable in dns or
         # fully qualified.
-        self.metadata['instance-id'] = md['instance-id']
+        self.metadata['instance-id'] = md.get('instance-id', serial)
```
Comment (smoser) [inline cloudinit/sources/DataSourceHetzner.py:78]: I wouldnt bother with this.
I'd rather check here that `md['instance-id'] == serial` and raise a runtime error.

But unless you a valid case where there may not be an instance-id in the metadata service (but there *is* a 'hostname' and we have an otherwise-functioning metadata service) then this is just confusing. and results in "sometimes i get the value from smbios and sometimes it comes from the MD".

Comment (asciiprod) [reply cloudinit/sources/DataSourceHetzner.py]: Good point. If both don't match, something is definitely wrong.

File: cloudinit/sources/DataSourceHetzner.py [hunk 5]
```diff
@@ -75,8 +81,18 @@ def get_data(self):
         self.metadata['public-keys'] = md.get('public-keys', None)
         self.vendordata_raw = md.get("vendor_data", None)
 
+        # instance-id and serial from SMBIOS should be identical
+        if self.metadata['instance-id'] != serial:
+            msg = "SMBIOS serial does not match instance ID from metadata"
+            LOG.critical(msg)
+            raise RuntimeError(msg)
```
Comment (smoser) [inline cloudinit/sources/DataSourceHetzner.py:88]: just one nit. just raise the exception, no need to log and raise. i know some places in cloud-init do that, but it really is unnecessary. cloud-init will fail if the exception is raised here (and it will get that message to the log). no additional copy of it is necessary.

File: cloudinit/sources/DataSourceHetzner.py [hunk 6]
```diff
@@ -96,8 +112,20 @@ def network_config(self):
...
+    if vendor_name != "Hetzner":
+        return (False, None)
+
+    serial = util.read_dmi_data("system-serial-number")
+    if serial:
+        LOG.debug("Running on Hetzner Cloud: serial=%s", serial)
+    else:
+        msg = "Hetzner Cloud detected, but no serial found"
+        LOG.critical(msg)
+        raise RuntimeError(msg)
```
Comment (smoser) [inline cloudinit/sources/DataSourceHetzner.py:126]: same comment.  just:

    raise RuntimeError("Hetzner Cloud detected, but no serial found")

**Review by smoser — CHANGES_REQUESTED** [review]
just two nits. then i'm happy with it.
thank you.

**Review by smoser — APPROVED** [review]

**Comment by smoser** [comment]
Just a comment, fix if you want.

```python
            msg = "SMBIOS serial does not match instance ID from metadata"
            raise RuntimeError(msg)
```

Is just as annoying as:

```python
            raise RuntimeError(
                "SMBIOS serial does not match instance ID from metadata"
```

but I think the second is slightly preferable just because it doesn't have the throw away variable.

**Comment by asciiprod** [comment]
> Just a comment, fix if you want.
> 
> ```python
>             msg = "SMBIOS serial does not match instance ID from metadata"
>             raise RuntimeError(msg)
> ```
> 
> Is just as annoying as:
> 
> ```python
>             raise RuntimeError(
>                 "SMBIOS serial does not match instance ID from metadata"
> ```
> 
> but I think the second is slightly preferable just because it doesn't have the throw away variable.

Agreed. Fixed

**Comment by TheRealFalcon** [comment]
Hi @asciiprod 

This change is included in the most recent cloud-init release, which will soon be included in the Ubuntu distributions Xenial, Bionic, Focal, and Groovy. Since we don't have access to the platform necessary to test this change, the cloud-init team is asking if you can test that this change works as expected in Ubuntu. We hope to be finished testing by the end of the week, so if you can't verify this functionality by then, we cannot guarantee that this change will work as expected in our Ubuntu releases. If you are able to test the feature, please add a comment to our [SRU process bug](https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/1905599) mentioning success or failure.

The steps to setup the proposed image are laid out [in the cloud-init docs](https://cloudinit.readthedocs.io/en/latest/topics/debugging.html?highlight=proposed#stable-release-updates-sru-testing-for-cloud-init).
